src/levanter/data/shardcache.py

```diff
@@ -0,0 +1,712 @@
+            self.current_chunk_writer = None
+
+            self._write_in_progress_metadata()
+            return current_metadata
+
+        return None
+
+    def _write_in_progress_metadata(self):
+        fs, path = fsspec.core.url_to_fs(self.cache_dir)
+        with fs.open(f"{path}/{self.shard_name}/metadata.json", "w") as f:
+            f.write(self._in_progress_metadata.to_json())
+
+    def finalize(self):
+        if self.current_chunk_writer is not None:
+            self.current_chunk_writer.close()
+            self.written_chunk_metadata.append(self.current_chunk_writer.metadata)
+            self.current_chunk_writer = None
+
+        try:
+            self.is_finished = True
+            self._write_in_progress_metadata()
+        except Exception:
+            logger.exception("Failed to finalize shard")
+            self.is_finished = False
+            raise
+
+    def _open_chunk_writer(self):
+        fs, path = fsspec.core.url_to_fs(self.cache_dir)
+        fs.makedirs(f"{path}/{self.shard_name}", exist_ok=True)
+
+        return _ChunkWriter(self.cache_dir, f"{self.shard_name}/chunk-{len(self.written_chunk_metadata)}")
+
+    def _load_metadata(self):
+        fs, path = fsspec.core.url_to_fs(self.cache_dir)
+        try:
+            with fs.open(f"{path}/{self.shard_name}/metadata.json", "r") as f:
+                obj = ShardMetadata.from_json(f.read())
+                self.written_chunk_metadata = obj.chunks
+                self.is_finished = obj.is_finished
+        except FileNotFoundError:
+            self.written_chunk_metadata = []
+            self.is_finished = False
+
+        self.current_chunk_writer = None
+
+
+def _seek_to_row(name, data_iterator, row_count, pos):
+    count = 0
+    for _ in tqdm(data_iterator, desc=f"[Shard {name}] seeking raw data iterator", total=row_count, position=pos + 1):
+        count += 1
+        if count >= row_count:
+            break
+
+
+def _take(source: Iterator[T], n: int) -> List[T]:
+    """Take the first n elements from an iterator, or until the iterator is exhausted."""
+    result = []
+    for _ in range(n):
+        try:
+            result.append(next(source))
+        except StopIteration:
+            break
+
+    return result
+
+
+def _exc_info():
+    exc_type, exc_value, exc_traceback = sys.exc_info()
+    tb = tblib.Traceback(exc_traceback)
+    return (exc_value, tb)
+
+
+def _restore_exc_info(exc_info):
+    exc_value, tb = exc_info
+    if exc_value is not None:
+        exc_value = exc_value.with_traceback(tb.as_traceback())
+    else:
+        return (None, None, tb.as_traceback())
+
+
+class _ShardState:
+    buffered_chunks: List[ChunkMetadata]
+    finished: bool
+
+    def __init__(self):
+        self.buffered_chunks = []
+        self.finished = False
+
+    def ready(self) -> bool:
+        return len(self.buffered_chunks) > 0 or self.finished
+
+
+@ray.remote
+class _ShardCacheManager(Generic[T]):
```

Comment:
We _may_ want actor fault tolerance, with would require us to either (1) save actor checkpoints or (2) make the actor be able to rebuild the chunk / shard state from the external disk state upon initalization. See [here](https://docs.ray.io/en/master/ray-core/actors.html#fault-tolerance) and [here](https://docs.ray.io/en/master/ray-core/fault_tolerance/actors.html#actor-fault-tolerance) in the Ray docs.

But it might also be simpler to just crash the whole job and restart it.